crates/ty_python_semantic/src/types.rs

```diff
@@ -5595,6 +5608,28 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Get the type of the `__class__` attribute of this type.
+    ///
+    /// For most types, this is equivalent to the meta type of this type. For `TypedDict` types,
+    /// this returns `type[dict[str, object]]` instead, because inhabitants of a `TypedDict` are
+    /// instances of `dict` at runtime.
+    #[must_use]
+    pub fn dunder_class(self, db: &'db dyn Db) -> Type<'db> {
+        if self.is_typed_dict() {
+            return SubclassOfType::from(
+                db,
+                KnownClass::Dict
+                    .to_specialized_class_type(
+                        db,
+                        [KnownClass::Str.to_instance(db), Type::object(db)],
+                    )
+                    .expect("dict[] takes two type parameters"),
+            );
+        }
```

Comment:
Shouldn't this be a class-literal type rather than a subclass-of type? Similar to the way we know that an int-literal must be an instance of exactly `int` (not a subclass of `int`), it's invalid for a `TypedDict` inhabitant to be an instance of a `dict` subclass -- it must be an instance of exactly `dict`, I think? This is different to `NominalInstance` types, which is why we use `SubclassOf` types as the meta-types for most instance types.